Fix ping example

jklmnn opened the pull request:
The ping example has been broken for quite a while. The current version uses an ICMP spec without a checksum aspect. Originally the checksum has been calculated after setting all fields (including the checksum field itself) and then replaced the original checksum. With the current version of PyRFLX setting a field while skipping verification leads to all following fields being invalidated. To generate a valid message these fields have to be set again with the same values.

treiher commented:
> The current version uses an ICMP spec without a checksum aspect.

Is there anything which prevents us from using an ICMP spec with checksum aspect in the ping example? IMHO the proposed fix does not make a good impression.

senier commented:
Is this related to #437?

jklmnn commented:
> Is this related to #437?

Yes, I forgot that #437 existed. It is intended to fix it.

jklmnn commented:
> > The current version uses an ICMP spec without a checksum aspect.
> 
> Is there anything which prevents us from using an ICMP spec with checksum aspect in the ping example? IMHO the proposed fix does not make a good impression.

I remember that we once had an ICMP spec with a checksum aspect, however it neither seems to be in the specifications repository nor in the specification directory we are using for this example.

treiher commented:
> I remember that we once had an ICMP spec with a checksum aspect, however it neither seems to be in the specifications repository nor in the specification directory we are using for this example.

I have found the following version:

```diff
diff --git a/icmp.rflx b/icmp.rflx
index f6d87cd..cc07ef7 100644
--- a/icmp.rflx
+++ b/icmp.rflx
@@ -109,7 +109,10 @@ package ICMP is
          Receive_Timestamp : Timestamp;
          Transmit_Timestamp : Timestamp
             then null;
-         Data : Opaque;
-      end message;
+         Data : Opaque
+            then null
+               if Checksum'Valid_Checksum;
+      end message
+         with Checksum => (Checksum => (Message'First .. Checksum'First - 1, Checksum'Last + 1 .. Message'Last));
 
 end ICMP;
```

Could you please check if that works with the ping example?

senier commented:
When in doubt, check with @rssen - he should have (or be able to come up with) the right spec as well as an implementation for the checksum.

jklmnn commented:
> > I remember that we once had an ICMP spec with a checksum aspect, however it neither seems to be in the specifications repository nor in the specification directory we are using for this example.
> 
> I have found the following version:
> 
> ```diff
> diff --git a/icmp.rflx b/icmp.rflx
> index f6d87cd..cc07ef7 100644
> --- a/icmp.rflx
> +++ b/icmp.rflx
> @@ -109,7 +109,10 @@ package ICMP is
>           Receive_Timestamp : Timestamp;
>           Transmit_Timestamp : Timestamp
>              then null;
> -         Data : Opaque;
> -      end message;
> +         Data : Opaque
> +            then null
> +               if Checksum'Valid_Checksum;
> +      end message
> +         with Checksum => (Checksum => (Message'First .. Checksum'First - 1, Checksum'Last + 1 .. Message'Last));
>  
>  end ICMP;
> ```
> 
> Could you please check if that works with the ping example?

I found two main issues. The proposed aspect doesn't work due to the improper handling of `Message'First` (see #503). If `Message'First` is replaced with `Tag'First` creating the IPv4 packet will fail since it then tries to parse it into a new message without a checksum function. For the message creation this could be fixed by allowing setting the icmp message directly (see #505) but for the parsing of the reply refinements need some information about checksum functions (see #504).

Since these changes seem to be non trivial I'd opt for the current solution to have a working example, as ugly as it is.

senier commented:
As discussed independently, we chose not to take the shortcut but fix the example by solving #503 and #504 (#505 is considered a convenience feature for now). SPARK code generation should be fine already, but we want to add a flag to `rflx` to enable treating the `Valid_Checksum` aspect as `True` in #507.

jklmnn commented:
I fixed the test including the issues #437, #503 and #504. @treiher @senier can you please check locally if the example works for you?

senier commented:
> I fixed the test including the issues #437, #503 and #504. @treiher @senier can you please check locally if the example works for you?

Works for me.

treiher commented:
> I fixed the test including the issues #437, #503 and #504. @treiher @senier can you please check locally if the example works for you?

The error seems to be fixed. But the output looks strange:

```Console
$ ./ping.py 127.0.0.1
PING 127.0.0.1 (127.0.0.1)
64 bytes from 127.0.0.1: icmp_seq=0
unexpected type
64 bytes from 127.0.0.1: icmp_seq=0
64 bytes from 127.0.0.1: icmp_seq=1
unexpected type
64 bytes from 127.0.0.1: icmp_seq=1
64 bytes from 127.0.0.1: icmp_seq=2
unexpected type
64 bytes from 127.0.0.1: icmp_seq=2
64 bytes from 127.0.0.1: icmp_seq=3
unexpected type
64 bytes from 127.0.0.1: icmp_seq=3
```

jklmnn commented:
> > I fixed the test including the issues #437, #503 and #504. @treiher @senier can you please check locally if the example works for you?
> 
> The error seems to be fixed. But the output looks strange:
> 
> ```
> $ ./ping.py 127.0.0.1
> PING 127.0.0.1 (127.0.0.1)
> 64 bytes from 127.0.0.1: icmp_seq=0
> unexpected type
> 64 bytes from 127.0.0.1: icmp_seq=0
> 64 bytes from 127.0.0.1: icmp_seq=1
> unexpected type
> 64 bytes from 127.0.0.1: icmp_seq=1
> 64 bytes from 127.0.0.1: icmp_seq=2
> unexpected type
> 64 bytes from 127.0.0.1: icmp_seq=2
> 64 bytes from 127.0.0.1: icmp_seq=3
> unexpected type
> 64 bytes from 127.0.0.1: icmp_seq=3
> ```

I fixed the behaviour in https://github.com/Componolit/RecordFlux/pull/501/commits/0be5ed518fb2d569a9792618529dd8f84b35ac90. Please check that it works for you.

treiher commented:
> I fixed the behaviour in [0be5ed5](https://github.com/Componolit/RecordFlux/commit/0be5ed518fb2d569a9792618529dd8f84b35ac90). Please check that it works for you.

It works for me now.

senier commented:
> > I fixed the behaviour in [0be5ed5](https://github.com/Componolit/RecordFlux/commit/0be5ed518fb2d569a9792618529dd8f84b35ac90). Please check that it works for you.
> 
> It works for me now.

Works for me, too.

jklmnn commented:
I also fixed the handling of invalid received packets in https://github.com/Componolit/RecordFlux/pull/501/commits/aa7af7b3870009a7ec93fcea3100315b92a46daa. Packets that are not valid echo replies are now dropped without sending an additional echo request.
The PR is now ready for a review again.

treiher commented on rflx/pyrflx/pyrflx.py:
```diff
@@ -18,13 +18,20 @@ def __init__(
-                self.__packages[p][str(m.name)] = MessageValue(
-                    m, model.refinements, skip_message_verification
-                )
+        messages: Dict[Tuple[str, str], MessageValue] = {}
+
+        for m in model.messages:
+            p = str(m.package)
+            if p not in self.__packages:
+                self.__packages[p] = Package(p)
+            message = MessageValue(m, skip_verification=skip_message_verification)
+            messages[(p, str(m.name))] = message
+            self.__packages[p][str(m.name)] = message
+
+        for r in model.refinements:
+            messages[(str(r.pdu.package.name), str(r.pdu.name))].add_refinement(
+                RefinementValue(r, messages[(str(r.sdu.package.name), str(r.sdu.name))])
+            )
```
Wouldn't it be easier to use `ID` instead of the tuple for `messages`?

jklmnn replied:
Fixed in https://github.com/Componolit/RecordFlux/pull/501/commits/1db3013a9f545ac919090b163a2d1a03cb4351af.
